src/api/v4/layer/layer.js

```diff
@@ -1,90 +1,274 @@
+  return this._featureClickColumns;
+};
+
+/**
+ * Set new columns for `featureOver` events
+ * 
+ * @param {Array<string>} columns An array containing column names
+ * @return {carto.layer.Layer} The layer
+ *
+ * @api
+ */
+Layer.prototype.setFeatureOverColumns = function (columns) {
+  this._featureOverColumns = columns;
+  if (this._internalModel) {
+    this._internalModel.tooltip.update(_getInteractivityFields(columns));
   }
-  return this._internalModel._engine.reload();
+
+  return this;
+};
+
+/**
+ * Return the columns available in `featureOver` events.
+ * 
+ * @return  {[Array<string>]} Column names available in `featureOver` events
+ *
+ * @api
+ */
+Layer.prototype.getFeatureOverColumns = function (columns) {
+  return this._featureOverColumns;
 };
 
+/**
+ * Hides the layer.
+ *
+ * @api
+ */
 Layer.prototype.hide = function () {
   this._visible = false;
   if (this._internalModel) {
-    this._internalModel.hide();
+    this._internalModel.set('visible', false);
   }
+
+  return this;
 };
 
+/**
+ * Shows the layer.
+ *
+ * @api
+ */
 Layer.prototype.show = function () {
   this._visible = true;
   if (this._internalModel) {
-    this._internalModel.show();
+    this._internalModel.set('visible', true);
   }
+
+  return this;
 };
 
-Layer.prototype.setSource = function (source) {
-  this._source = source;
-  if (this._internalModel) {
-    this._internalModel.set('source', source);
+// Private functions.
+
+Layer.prototype._createInternalModel = function (engine) {
+  return new CartoDBLayer({
+    id: this._id,
+    source: this._source.$getInternalModel(),
+    cartocss: this._style.toCartoCSS(),
+    visible: this._visible,
+    infowindow: _getInteractivityFields(this._featureClickColumns),
+    tooltip: _getInteractivityFields(this._featureOverColumns)
+  }, { engine: engine });
+};
+
+Layer.prototype._reloadEngine = function () {
+  if (this._engine) {
+    return this._engine.reload();
   }
+
+  return Promise.resolve();
 };
 
+// Internal functions.
+
 Layer.prototype.$setEngine = function (engine) {
+  if (this._engine) {
+    return;
+  }
+  this._engine = engine;
   this._source.$setEngine(engine);
   if (!this._internalModel) {
-    this._internalModel = new CartoDBLayer({
-      id: this._id,
-      source: this._source.$getInternalModel(),
-      cartocss: this._style.toCartoCSS(),
-      visible: this._visible
-    }, { engine: engine });
+    this._internalModel = this._createInternalModel(engine);
   }
 };
 
-Layer.prototype.$getInternalModel = function () {
-  return this._internalModel;
-};
+// Scope functions
+
+/**
+ * Transform the columns array into the format expected by the CartoDBLayer.
+ */
+function _getInteractivityFields (columns) {
+  var fields = columns.map(function (column, index) {
+    return {
+      name: column,
+      title: true,
+      position: index
+    };
+  });
+
+  return {
+    fields: fields
+  };
+}
+
+function _checkStyle (style) {
```

Comment:
should we move the checks to the Base class to be consistent with the dataviews?